### ISSUE TITLE
Yet another request for equation reference

### ISSUE THREAD

**Issue by wujinq** [KaTeX/KaTeX#2798]
> Before requesting a feature, please search for [existing issues](https://github.com/KaTeX/KaTeX/issues).

**Is your feature request related to a problem? Please describe.**
Equation labeling and reference (`\label` and `\eqref`) are handy features in technical and scientific writing. The features have been requested in issue #350, #571 and #2003, but little further progress seems to be made. 

**Describe the solution you'd like:**

The full `\label` and `\eqref` requires three things:

- `\tag` to create equation tags, which is already implemented
- `\label` to give a label to an equation, and to automatically add a `\tag` to that equation
- `\eqref` to refer to the equation

A possible reason that the last two features are not implemented may be that what exactly the number `\eqref` displays cannot be determined by a single scan of the page: in LaTeX you need to compile the file _twice_ to get everything done, causing performance overhead. (I'm not familiar with the inner implementation of KaTeX so sorry for anything inaccurate in my comments.)

A more practical solution may be to implement `\label` first: KaTeX keeps a global counter, and when KaTeX meets a math block with `\label{the-label}`, the counter +1 and a `\tag` is inserted, and an anchor with an id like `katex-eq-the-label` is also inserted. 

`\eqref` can then be implemented in a progressive enhancement manner:
1. The costless approach is to forget about `\eqref` and let users just use a hyperlink with the target `#katex-eq-the-label` and specify the hyperlink text by themselves. In this way equations themselves are automatically numbered, but users still have to check the number of an equation when refering to that equation.
2. A slight enhancement is to include `\eqref` in a very restricted manner: for example, `\eqref` is included but the only legit way to use it is `$\eqref{the-label}$` with nothing else in the block. KaTeX ignores these `\eqref`s when displaying math, and does a fast second pass to find all `$\eqref{the-label}$` and replace them with hyperlinks whose targets are the labels and whose hyperlink texts are the numbers of equations. In this way the performance may still be acceptable because the second pass doesn't start the whole math engine. This should cover most of the equation reference demands.
3. Maybe one day the full `\eqref` is implemented. 

**Describe alternatives you've considered:**
Create a markdown-it plugin that does the same work; this is outside the discussion about KaTeX itself, though.

**Comment by ronkok** [comment]
@wujinq You are correct in that `\ref` requires a pass through the entire document.

As it happens, I've already implemented `\label` and `\ref` in my new library, [Temml](https://temml.org/). I could back-port the work from there into KaTeX. Before I start work, I'd like to hear from others what they think about how the implementation works.

In order to support `\label` and `\ref`, I exposed a new function, `postProcess()` . It does the pass through the entire document noted above. A website administrator who wanted to support `\ref` would have to make a call to `katex.postProcess()` after all the calls to `katex.render` are complete. The source code in a web page might look something like this:

```
const macros = {};
for (let element of mathElements) {
    katex.render(element.textContent, element, { macros });
}
katex.postProcess(document.body);
```

For the other contributors here, please let me know what you think of that approach. If it is acceptable, I can do the back-port.

**Comment by ronkok** [comment]
For pages whose math has been rendered server-side, the `postProcess()` function would still run client-side, if at all. There would be a small client-side file, `katexPostProcess.js`. It would contain nothing but the `postProcess()` function.

**Comment by kevinbarabash** [comment]
Is it safe to assume that `\ref` can reference equations that haven't been rendered yet?  If so, I don't see any other way than to run some sort of post-processing.  How does `postProcess()` know what to do when looking at the DOM?  Do it grab the source annotation from the DOM nodes?

**Comment by ronkok** [comment]
> \ref can reference equations that haven't been rendered yet?

Yes.

> How does `postProcess()` know what to do when looking at the DOM? 

When tags or AMS environments are created, a breadcrumb is left behind in the form a CSS class, just for this purpose. Then `postProcess()` can find the relevant HTML elements with just a call to `document.getElementsByClassName()`.

The `postProcess()` function itself is pretty easy. There is more work involved in leaving behind the breadcrumbs.

**Comment by N0rbert** [comment]
The following syntax for [xaringan](https://bookdown.org/yihui/rmarkdown/xaringan.html) is needed:

```
\begin{equation}
  g\left(k\right) = \binom{n}{k} p^k\left(1-p\right)^{n-k}
\label{binom2}
\end{equation}
```

Currently it is not rendered:

![label-fail](https://user-images.githubusercontent.com/309456/110846275-a8a85a00-82bc-11eb-9c37-2bafaaa3a97d.png)

> KaTeX parse error: Undefined control sequence: \label at position 77: …p\right)^{n-k} \̲l̲a̲b̲e̲l̲{binom2} \end{e…

I hope you implement it.

**Comment by wujinq** [comment]
I found [a VSCode extension](https://marketplace.visualstudio.com/items?itemName=mathpix.vscode-mathpix-markdown) which behaves exactly as I wish, except the `\eqref` is implemented via the plain old HTML link. I'm not familiar with how to make VSCode extensions  but the implementation of the project doesn't seem complicated. Maybe it is a good idea to reuse this extension?

**Comment by edemaine** [comment]
> For the other contributors here, please let me know what you think of that approach. If it is acceptable, I can do the back-port.

@ronkok Sorry for taking so long to think about this. Now having some experience with the new equation numbering features in [my project](https://github.com/edemaine/coauthor/), I'm excited by this approach, but I have some complications to add.

In my setting, the HTML document contains multiple messages each of which is treated as its own LaTeX document (e.g. macros are local to the document). I think other systems behave similarly, with multiple "LaTeX contexts" possibly on the same page. (I think I wrote a PR for Gitlab that did this too, though I forget whether this got merged.)

I imagine `\label`s should be local to each context. This means a couple of things:
1. Shouldn't use the `id` attribute, because that needs to be globally unique. I didn't look at what Temml uses, but we can easily use a custom attribute like `data-katex-label`.
2. `katex.postProcess` should take an `HTMLElement` argument, similar to `renderMathInElement`, so that it can be scoped to a subtree of `document.body`.

Hopefully these aren't difficult constraints to work with?

**Comment by N0rbert** [comment]
```
\begin{equation}
  g\left(k\right) = \binom{n}{k} p^k\left(1-p\right)^{n-k}
\label{binom2}
\end{equation}
```

still does not work.

**Comment by N0rbert** [comment]
The following equation

```
\begin{equation}
  g\left(k\right) = \binom{n}{k} p^k\left(1-p\right)^{n-k}
\label{binom2}
\end{equation}
```

still does not render with latest KaTeX 0.13.16 - "KaTeX parse error: Undefined control sequence: \label at position 77: …p\right)^{n-k} \̲l̲a̲b̲e̲l̲{binom2} \end{e…".

**Comment by N0rbert** [comment]
The following equation
```
\begin{equation}
  g\left(k\right) = \binom{n}{k} p^k\left(1-p\right)^{n-k}
\label{binom2}
\end{equation}
```

still does not render with latest KaTeX 0.13.16 - "KaTeX parse error: Undefined control sequence: \label at position 77: …p\right)^{n-k} \̲l̲a̲b̲e̲l̲{binom2} \end{e…".

**Comment by edemaine** [comment]
That's correct: `\label` hasn't been implemented yet. This issue (and #2003) tracks the progress.

**Comment by N0rbert** [comment]
The following equation
```
\begin{equation}
  g\left(k\right) = \binom{n}{k} p^k\left(1-p\right)^{n-k}
\label{binom2}
\end{equation}
```

still does not render with latest KaTeX 0.16.9 - "KaTeX parse error: Undefined control sequence: \label at position 77: …p\right)^{n-k} \̲l̲a̲b̲e̲l̲{binom2} \end{e…".
